lua/wikis/commons/Infobox/Extension/TeamHistoryAuto.lua

```diff
@@ -0,0 +1,501 @@
+local POSITION_ICON_DATA = Lua.requireIfExists('Module:PositionIcon/data', {loadData = true})
+
+---@class TeamHistoryAuto
+---@operator call(table?): TeamHistoryAuto
+---@field config {player: string, hasHeaderAndRefs: boolean?}
+---@field transferList table[]
+local TeamHistoryAuto = Class.new(function(self, args)
+	---@type {player: string?}
+	args = args or {}
+	local configFromInfo = (Info.config.infoboxPlayer or {}).automatedHistory or {}
+	self.config = {
+		player = (args.player or mw.title.getCurrentTitle().subpageText):gsub('^%l', string.upper),
+		hasHeaderAndRefs = configFromInfo.hasHeaderAndRefs,
+	}
+end)
+
+---@return self
+function TeamHistoryAuto:store()
+	if not Namespace.isMain() then return self end
+	Array.forEach(self.transferList, function(transfer, transferIndex)
+		self:_checkForMissingLeaveDate(transfer, transferIndex)
+		local teamLink = self:_getTeamLinkAndText(transfer)
+		if not teamLink and not transfer.role then return end
+
+		mw.ext.LiquipediaDB.lpdb_datapoint('Team_'..transferIndex, {
+			type = 'teamhistory',
+			name = self.config.player,
+			information = teamLink,
+			extradata = mw.ext.LiquipediaDB.lpdb_create_json({
+				joindate = transfer.joinDate,
+				leavedate = transfer.leaveDate or '2999-01-01',
+				teamcount = transferIndex,
+				role = transfer.role,
+				auto = 1,
+			})
+		})
+	end)
+
+	return self
+end
+
+---@param transfer table
+---@param transferIndex integer
+function TeamHistoryAuto:_checkForMissingLeaveDate(transfer, transferIndex)
+	if transferIndex == #self.transferList or transfer.leaveDate then return end
+	mw.ext.TeamLiquidIntegration.add_category('Players with potential incomplete transfer history')
+end
+
+---@param transfer table
+---@return string?
+---@return Widget
+function TeamHistoryAuto:_getTeamLinkAndText(transfer)
+	if Logic.isEmpty(transfer.team) and Table.includes(SPECIAL_ROLES_LOWER, (transfer.role or ''):lower()) then
+		return nil, HtmlWidgets.B{children = {transfer.role}}
+	elseif not mw.ext.TeamTemplate.teamexists(transfer.team) then
+		return transfer.team, Link{link = transfer.team}
+	end
+	local leaveDateCleaned = TeamHistoryAuto._adjustDate(transfer.leaveDate)
+	local teamData = mw.ext.TeamTemplate.raw(transfer.team, leaveDateCleaned) or {}
+
+	return teamData.page, Link{
+		link = teamData.page,
+		children = {TeamHistoryAuto._shortenTeamName(teamData)}
+	}
+end
+
+---@param teamData {name: string, bracketname: string, shortname: string}
+---@return string
+function TeamHistoryAuto._shortenTeamName(teamData)
+	if string.len(teamData.name) <= 17 then
+		return teamData.name
+	elseif string.len(teamData.bracketname) <= 17 then
+		return teamData.bracketname
+	else
+		return teamData.shortname or teamData.name
+	end
+end
+
+-- earlier date for fromteam to account for rebrands
+---@param date string?
+---@return string?
+function TeamHistoryAuto._adjustDate(date)
+	if type(date) ~= 'string' or Logic.isEmpty(date) then
```

Comment:
When can date be a non-string?

Reply:
if it is nil :P (but that would be catched by Logic.isEmpty